packages/runtime/container-runtime/src/opLifecycle/opGroupingManager.ts

```diff
@@ -0,0 +1,71 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { ContainerRuntimeMessage } from "..";
+import { IBatch } from "./definitions";
+
+interface IGroupedMessage {
+	contents: string;
+	metadata: Record<string, unknown> | undefined;
+	compression: string;
+}
+
+export class OpGroupingManager {
+	static groupedBatchOp = "groupedBatch";
+
+	constructor(private readonly groupedBatchingEnabled: boolean) {}
+
+	public groupBatch(batch: IBatch): IBatch {
+		if (batch.content.length < 2 || !this.groupedBatchingEnabled) {
+			return batch;
+		}
+
+		// Need deserializedContent for back-compat
+		const deserializedContent = {
+			type: OpGroupingManager.groupedBatchOp,
+			contents: batch.content.map((message) => ({
+				contents: message.contents,
+				metadata: message.metadata,
+				compression: message.compression,
+			})),
+		};
+
+		const groupedBatch: IBatch = {
+			...batch,
+			content: [
+				{
+					...batch.content[0],
+					metadata: undefined, // important to not keep reference to first op's batch metadata
+					compression: undefined,
+					deserializedContent: deserializedContent as ContainerRuntimeMessage,
+					contents: JSON.stringify(deserializedContent),
+				},
+			],
+		};
+		return groupedBatch;
+	}
+
+	public ungroupOp(op: ISequencedDocumentMessage): ISequencedDocumentMessage[] {
+		if (
+			!op.contents ||
+			typeof op.contents !== "object" ||
```

Comment:
null is an object. I suggest not to test for it (this line and line above), but change next line to 

op?.contents?.type !== OpGroupingManager.groupedBatchOp

Basically, in JS, everything is an object. So either we can get to the deep property, or test fails sooner.